src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -211,7 +211,7 @@ public override BoundStatement InstrumentReturnStatement(BoundReturnStatement or
 
             // A synthesized return statement that does not return a value never requires instrumentation.
             // A property set method defined without a block has such a synthesized return statement.
-            if (!_methodHasExplicitBlock && ((BoundReturnStatement)original).ExpressionOpt != null)
+            if ((!_methodHasExplicitBlock || _methodBodyFactory.CurrentMethod.MethodKind == MethodKind.LambdaMethod) && ((BoundReturnStatement)original).ExpressionOpt != null)
```

Comment:
What is the syntax associated with a compiler-synthesized return statement/expression? In the situation of interest, the return statement is synthesized by the compiler. If the syntax is something reliable, from which one can get to the enclosing lambda, then great. (My guess is that the syntax of the returned expression, and probably that of the return statement itself, would be that of the expression that is the body of the lambda, but I'd like confirmation.)